### ISSUE TITLE
Update for new versions of SQLAlchemy

### ISSUE THREAD

**Issue by davidism** [pallets/flask-sqlalchemy#910]
SQLAlchemy 1.4 made some changes that are currently incompatible with this extension. Identify failing tests and make changes to fix those. We should also pin to `SQLAlchemy<2`, since that will be a bigger change and will be more appropriate in Flask-SQLAlchemy 3.0.

Here's some that I'm aware of so far:

- [x] #885 `URL` is immutable

**Comment by lensonp** [comment]
Possibly related. Test suites started failing on `db.drop_all()` after sqlalchemy 1.3.23->1.4.0.

flask-sqlalchemy 2.4.4
sqlalchemy 1.4.0
pymysql 1.0.2
mysql  Ver 8.0.23-0ubuntu0.20.04.1 for Linux on x86_64 ((Ubuntu))
Python 3.8.5
Flask 1.1.2

Stack trace:
```
tests/test_all.py:23: in <module>
    db.drop_all()
/usr/local/lib/python3.6/dist-packages/flask_sqlalchemy/__init__.py:1047: in drop_all
    self._execute_for_all_tables(app, bind, 'drop_all')
/usr/local/lib/python3.6/dist-packages/flask_sqlalchemy/__init__.py:1031: in _execute_for_all_tables
    op(bind=self.get_engine(app, bind), **extra)
/usr/local/lib/python3.6/dist-packages/flask_sqlalchemy/__init__.py:962: in get_engine
    return connector.get_engine()
/usr/local/lib/python3.6/dist-packages/flask_sqlalchemy/__init__.py:555: in get_engine
    options = self.get_options(sa_url, echo)
/usr/local/lib/python3.6/dist-packages/flask_sqlalchemy/__init__.py:570: in get_options
    self._sa.apply_driver_hacks(self._app, sa_url, options)
/usr/local/lib/python3.6/dist-packages/flask_sqlalchemy/__init__.py:884: in apply_driver_hacks
    sa_url.query.setdefault('charset', 'utf8')
E   AttributeError: 'sqlalchemy.cimmutabledict.immutabledict' object has no attribute 'setdefault'
```

**Comment by lensonp** [comment]
I'm not sure, but I suspect the above error is related to one of the following column definitions:
```
deferred(db.Column(MutableDict.as_mutable(db.JSON), default={}))
deferred(db.Column(db.JSON(), default={}))
deferred(db.Column(db.JSON()))
```

**Comment by dpgaspar** [comment]
Finding issues also:

``` python
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/Flask_AppBuilder-3.2.0-py3.7.egg/flask_appbuilder/security/sqla/manager.py", line 88, in create_db
    engine = self.get_session.get_bind(mapper=None, clause=None)
  File "<string>", line 2, in get_bind
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/sqlalchemy/orm/scoping.py", line 104, in _proxied
    return self.registry()
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/sqlalchemy/util/_collections.py", line 1010, in __call__
    return self.registry.setdefault(key, self.createfunc())
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/sqlalchemy/orm/session.py", line 4027, in __call__
    return self.class_(**local_kw)
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py", line 138, in __init__
    bind = options.pop('bind', None) or db.engine
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py", line 943, in engine
    return self.get_engine()
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py", line 962, in get_engine
    return connector.get_engine()
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py", line 555, in get_engine
    options = self.get_options(sa_url, echo)
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py", line 570, in get_options
    self._sa.apply_driver_hacks(self._app, sa_url, options)
  File "/Users/daniel/workarea/preset/Flask-AppBuilder/venv/lib/python3.7/site-packages/flask_sqlalchemy/__init__.py", line 914, in apply_driver_hacks
    sa_url.database = os.path.join(app.root_path, sa_url.database)
AttributeError: can't set attribute
```

**Comment by ThiefMaster** [comment]
Just FYI, I would strongly suggest anyone to pin sqlalchemy to <1.4 - 1.x release in SA are generally major/breaking releases.

**Comment by brmzkw** [comment]
This is breaking with SQLAlchemy 1.4:

```python
from flask import Flask
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy.orm import joinedload

app = Flask(__name__)

app.config['SQLALCHEMY_DATABASE_URI'] = 'sqlite:///:memory:'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)

class User(db.Model):
    id = db.Column(db.Integer, primary_key=True)
    manager_id = db.Column(db.Integer, db.ForeignKey('user.id'))
    manager = db.relationship('User', remote_side=[id])


db.create_all()

u = User()
db.session.add(u)
db.session.commit()

u2 = User()
db.session.add(u2)
db.session.commit()

paginate = db.session.query(User, User.manager).options(joinedload(User.manager)).paginate()
```

**Comment by ThiefMaster** [comment]
Including errors/tracebacks would be useful, and if it's just for people googling their error to find this issue...

**Comment by zzzeek** [comment]
for the URL stuff please note [url is now immutable](https://docs.sqlalchemy.org/en/14/changelog/migration_14.html#the-url-object-is-now-immutable),   there was unfortunately no reasonable deprecation path for this combined with the fact that I didn't think mutation of URLs was that prevalent anyway.   if flask is trying to "fix up" the arguments when the engine is created it would better if they used the [event hook](https://docs.sqlalchemy.org/en/14/core/engines.html#controlling-how-parameters-are-passed-to-the-dbapi-connect-function) to do so, that way you get the arguments that are going straight to dbapi.connect() and you can deal with sqlite3 or whatever directly there.  `dialect.name` will tell you what kind of database you're dealing with and ``dialect.driver`` will give you the DBAPI name.

**Comment by zzzeek** [comment]
Here's a trace for part of flask admin, separate project?     they should not be accessing private API like that, if they need class regsitry access it would have been nice if we had been tasked to create a public API 

```
  File "/usr/local/bin/airflow", line 26, in <module>
    from airflow.bin.cli import CLIFactory
  File "/usr/local/lib/python3.7/dist-packages/airflow/bin/cli.py", line 80, in <module>
    from airflow.www.app import (cached_app, create_app)
  File "/usr/local/lib/python3.7/dist-packages/airflow/www/app.py", line 38, in <module>
    from airflow.www.blueprints import routes
  File "/usr/local/lib/python3.7/dist-packages/airflow/www/blueprints.py", line 25, in <module>
    from airflow.www import utils as wwwutils
  File "/usr/local/lib/python3.7/dist-packages/airflow/www/utils.py", line 40, in <module>
    import flask_admin.contrib.sqla.filters as sqlafilters
  File "/usr/local/lib/python3.7/dist-packages/flask_admin/contrib/sqla/__init__.py", line 2, in <module>
    from .view import ModelView
  File "/usr/local/lib/python3.7/dist-packages/flask_admin/contrib/sqla/view.py", line 18, in <module>
    from flask_admin.contrib.sqla.tools import is_relationship
  File "/usr/local/lib/python3.7/dist-packages/flask_admin/contrib/sqla/tools.py", line 4, in <module>
    from sqlalchemy.ext.declarative.clsregistry import _class_resolver
ModuleNotFoundError: No module named 'sqlalchemy.ext.declarative.clsregistry'
```

**Comment by ThiefMaster** [comment]
yes, separate project: https://github.com/flask-admin/flask-admin

**Comment by davidism** [comment]
Fixed the URL issue in #933, as far as I can tell that's the only incompatibility. Will release 2.5 soon.
